### PR TITLE
Use the new cloud centric storage to fetch/store results running ert3

### LINKED ISSUE #1461: Store records and ensembles in psql-backed ert-storage when running ERT3
Aiming to use https://github.com/equinor/ert-storage as the storage solution, replacing the exisitng yaml-storage.

### PR DESCRIPTION
**Pull request by ManInFez**
resolves #1461 

Splitting github actions in to macos and linux, since macos workers does not support services. Service is needed for having a pgSQL server available. 

Ert-storage is installed by cloning git, since alembic needs the alembic.ini file for migrating the server. The migrations should in ert-storage should be moved in to the code,  in the same manner as  new-legacy storage.

All record data is  at this time is stored as binary files. no interpretation or structure is read on the ert-storage side. Support for float vectors are available in ert storage and when the configuration is ready, we should be able use them :)

### CODE REVIEW

File: .github/workflows/build-linux.yml [hunk 1]
```diff
@@ -45,12 +67,7 @@ jobs:
         sleep 3
         herbstluftwm &
         sleep 1
-        python -m pytest -sv
-
-    - name: Test MacOS
-      if: matrix.os == 'macos-latest'
-      run: |
-        python -m pytest -sv
+        python -m pytest -sv -k "not ert3"
```
Comment (jondequinor) [inline .github/workflows/build-linux.yml:70]: Could you give some context here?

File: ert3/engine/_run.py [hunk 2]
```diff
@@ -75,7 +68,7 @@ def _load_ensemble_parameters(ensemble, workspace):
 def _prepare_sensitivity(ensemble, workspace_root, experiment_name):
     parameter_distributions = _load_ensemble_parameters(ensemble, workspace_root)
     input_records = ert3.algorithms.one_at_the_time(parameter_distributions)
-
+    print("prepare sens:", input_records)
```
Comment (DanSava) [inline ert3/engine/_run.py:71]: do you need this ?

File: ert3/data/_record.py [hunk 3]
```diff
@@ -58,6 +59,7 @@ class MultiEnsembleRecord(_DataElement):
     def __init__(
         self, *, ensemble_records, record_names=None, ensemble_size=None, **kwargs
     ):
+        print(ensemble_records)
```
Comment (DanSava) [inline ert3/data/_record.py:62]: is this required?

File: tests/ert3/storage/test_storage.py [hunk 4]
```diff
@@ -149,7 +162,12 @@ def test_get_record_names(tmpdir):
...
             experiment_records[str(experiment)].append(name)
 
-            recnames = ert3.storage.get_ensemble_record_names(
+            ens_record_names = ert3.storage.get_ensemble_record_names(
                 workspace=tmpdir, experiment_name=experiment
             )
-            assert sorted(experiment_records[str(experiment)]) == sorted(recnames)
+            print(ens_record_names)
+
+            print(experiment_records)
```
Comment (DanSava) [inline tests/ert3/storage/test_storage.py:170]: is this needed ?

File: .github/workflows/build-linux.yml (hunk 1 above)
Comment (ManInFez) [reply .github/workflows/build-linux.yml:70]: Ert3 is now running against ert-storage. Ert-storage needs a pgsql instance, and services on github actions are only available for linux workers.

File: ert3/engine/_run.py (hunk 2 above)
Comment (ManInFez) [reply ert3/engine/_run.py:71]: need is a strong word..

File: ert3/data/_record.py (hunk 3 above)
Comment (jondequinor) [inline ert3/data/_record.py:62]:
```suggestion

```

File: ert3/engine/_run.py (hunk 2 above)
Comment (jondequinor) [inline ert3/engine/_run.py:71]:
```suggestion

```

File: ert3/engine/_record.py [hunk 5]
```diff
@@ -37,4 +37,5 @@ def sample_record(
         record_name=record_name,
         ensemble_record=ensrecord,
         experiment_name=experiment_name,
+        record_class="parameter",
```
Comment (jondequinor) [inline ert3/engine/_record.py:40]: I think we should have the record class on the record itself. So pass it to `EnsembleRecord` a few lines up.

File: ert3/storage/_storage.py [hunk 6]
```diff
@@ -70,7 +81,7 @@ def get_experiment_names(*, workspace):
     return experiment_names
 
 
-def _add_data(workspace, experiment_name, data_type, data):
+def _add_data(workspace, experiment_name, record_name, data, record_class="Normal"):
```
Comment (jondequinor) [inline ert3/storage/_storage.py:84]: Move it to the record proper

**Comment by jondequinor** [comment]
We've been discussing the implications of ert-storage being postgres-only, and have some questions. This isn't really about this PR, though, but this PR will be affected by the outcome of this discussion.

- Lack of a local (non-postgres) storage solution implementation implies
  - running ERT in any way requires a running postgres server
  - integration testing locally requires you to run a postgres server
  - fmu development of models will need to use **central storage**, even though no compute cluster is involved\*
  - CI, and local testing without postgres differs as CI will run postgres and thus more tests.
- We probably want to mark storage integration testing as exactly that, and then run them _after_ unit tests. This is good, because integration testing should ideally test a superset of the unit tests.

\* _having the researcher run postgres locally on rgs is not feasible_

I've seen https://github.com/equinor/ert-storage/wiki/Running-Postgres-with-user-privileges but `initdb` isn't available on rgs, is there something I need to source?

We've decided to defer until @markusdregi has looked at this as well.

File: .github/workflows/build-linux.yml (hunk 1 above)
Comment (jondequinor) [reply .github/workflows/build-linux.yml:70]: I mean, as a comment in the code

**Comment by dotfloat** [comment]
**Lack of local (non-postgres) storage solution implementation implies**

> running ERT in any way requires a running postgres server (even for workflows, if they employ records, which they probably will)

Who is this ominous "they"? If it's assets within Equinor, then they'll be using a postgres. You can have a user postgres server, and we can have a script that manages your local  development server on RGS.

Students et al. can install Postgres on their own machines and have it work.

> integration testing locally requires you to run a postgres server

Yes. I think it's reasonable that as developers we're able to set up a local postgres server. In the (near) future, we'll have a `docker-compose.yml` that will give you a testing environment.

> FMU development of models will need to use **central storage**, even though no compute cluster is involved*. This is potentially not a problem if we have a "scratch" postgres.
> \* _having the researcher run postgres locally on rgs is not feasible_

Why is it not feasible?

> CI, and local testing without postgres differs as CI will run postgres and thus more tests.

I think this is a reasonable compromise.

> We probably want to mark storage integration testing as exactly that, and then run them after unit tests. This is good, because integration testing should ideally test a superset of the unit tests.

👍 

> I've seen https://github.com/equinor/ert-storage/wiki/Running-Postgres-with-user-privileges but `initdb` isn't available on rgs, is there something I need to source?

It's something we need to ask S@E to install I guess.

**Comment by ManInFez** [comment]
**From todays discussion on ert3 with ert-storage** 
- new ert (cloud) storage should get local capabilities, meaning that one ca run the ert-storage module without requiring pgSQL and other big services (radis blob store) etc. 
- The local solution is to be, in the first iteration, backed by SQLite, and should not support migrations.
- The local solution should be developed before merging the current ert3-storage PR
- Original ERT3-tests should be run on MacOS and linux using local solution
- New CI tests should be generated using pgSQL (and other services in the future)

**Comment by markusdregi** [comment]
Looks like a good start 👍 I was kind of hoping for this PR to not touch the `ert3` code besides the `ert3.storage` module. Could you motivate the extension and change of the API of `ert3.storage`?

**Comment by ManInFez** [comment]
> Looks like a good start 👍 I was kind of hoping for this PR to not touch the `ert3` code besides the `ert3.storage` module. Could you motivate the extension and change of the API of `ert3.storage`?

With the new version of ert-storage, records are now connected to ensembles either as inputs (old word for parameter) or outputs. This requires that the storage module needs to know what class it is when pushing a record to storage. So it was either to change the  'add_ensemble_record' signature or use extend the storage api with 'add_ensemble_input', 'add_ensemble_outut'. The former was chosen as it suggested the minimal change outside the storage module. For exporting, fetching inputs and outputs from an ensemble was more clean and faster using a single call for each, as fetching a list, and then fetching each record separately seemed a bit cumbersome.

**Comment by markusdregi** [comment]
> For exporting, fetching inputs and outputs from an ensemble was more clean and faster using a single call for each, as fetching a list, and then fetching each record separately seemed a bit cumbersome.

I can agree to that testament with the current design where the client is responsible for all the data passing with storage, but we are in the process of moving this responsibility over to the compute side when running forward models. Which implies that most input fetching and output publishing will be done on the compute side and then we really need the granularity to pass a record for a single realisation...

> With the new version of ert-storage, records are now connected to ensembles either as inputs (new word for parameter) or outputs. This requires that the storage module needs to know what class it is when pushing a record to storage. So it was either to change the  'add_ensemble_record' signature or use extend the storage api with 'add_ensemble_input', 'add_ensemble_outut'.

But each experiment is passed the names of its parameters upon creation. Should that not suffice to deduce whether a record is input or output? I just got rid of the `add_input` and `add_output` methods in `ert3.storage` because I don't think the record itself should consider itself either input or output nor the application at large. Because we will soon end up in the scenario where one stage's (or pipeline or iteration or forward model or step) output is another stage's input. Hence, the perspective of whether some data is input or output should reside within the consuming entity, not within the data itself. At least from a workflow perspective. Hence, I would prefer if we can avoid leaking that information into all of `ert3`...

**Comment by ManInFez** [comment]
So in the database the actual records are connected to the ensemble as inputs and outputs, not a list of strings. There is an actual relation in the database, which means that you would have to assign them in one way or the other. We of course just use a loose relation with a social contract that this particular string list should be the inputs. 

All ensembles/workflows/stages will have some form of inputs and outputs, and I think it makes sense to have the information of who produced it as well as who consumed it.

**Comment by ManInFez** [comment]
We do allow for passing single realization-records both ways.

**Comment by jondequinor** [comment]
> > running ERT in any way requires a running postgres server (even for workflows, if they employ records, which they probably will)
> 
> Who is this ominous "they"? If it's assets within Equinor, then they'll be using a postgres. You can have a user postgres server, and we can have a script that manages your local development server on RGS.

It's the workflows that are the _they_.

> > * _having the researcher run postgres locally on rgs is not feasible_
> 
> Why is it not feasible?

- Complexities associated with indirectly being a pgsql rdbms admin in _the long run_.
- Do you do `initdb` on `/private` or `/tmp`? Both have pros and cons, but the former is probably lunacy.
- Local development and sqlite is a great combo.

**Comment by ManInFez** [comment]
New pr makes this one irrelevant https://github.com/equinor/ert/pull/1490

**Pull request closed without merging**